### PR TITLE
Lock free queue

### PR DESCRIPTION
**Pull request by tsvisabo**
- Lock free queue wrapper for libdfs umm queue

### CODE REVIEW

File: utils/containers/lock_free/lf_queue.c.tpl [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+void iota_lock_free_queue_umm_init_user(iota_lf_umm_queue_{TYPE}_t* const queue) {
+  LFDS711_MISC_MAKE_VALID_ON_CURRENT_LOGICAL_CORE_INITS_COMPLETED_BEFORE_NOW_ON_ANY_OTHER_LOGICAL_CORE;
+}
+
+void* iota_lock_free_queue_umm_free(iota_lf_umm_queue_{TYPE}_t* const queue) {
+  if (queue == NULL) {
+    return;
+  }
+  lfds711_queue_umm_cleanup(&queue->queue, NULL);
```
Comment (th0br0) [inline utils/containers/lock_free/lf_queue.c.tpl:35]: What happens to the elements not in the freelist but inside the queue? Shouldn't those be added to the freelist on cleanup iteration?

File: utils/containers/lock_free/lf_queue.c.tpl [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+retcode_t iota_lock_free_queue_umm_enqueue(iota_lf_umm_queue_{TYPE}_t* const queue,
+                                           void const* const data) {
+  iota_lf_queue_umm_{TYPE}_t* p = NULL;
+
+  if (queue == NULL || data == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  struct lfds711_freelist_element* fe;
```
Comment (th0br0) [inline utils/containers/lock_free/lf_queue.c.tpl:54]: could we move this declaraiton to the start of the method?

File: utils/containers/lock_free/lf_queue.c.tpl [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+void* iota_lock_free_queue_umm_free(iota_lf_umm_queue_{TYPE}_t* const queue) {
+  if (queue == NULL) {
+    return;
+  }
+  lfds711_queue_umm_cleanup(&queue->queue, NULL);
+
+  struct lfds711_freelist_element* fe;
+  iota_lf_queue_umm_{TYPE}_t* queue_element;
```
Comment (th0br0) [inline utils/containers/lock_free/lf_queue.c.tpl]: could we move these declarations to the start of the method?

File: utils/containers/lock_free/lf_queue.c.tpl [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    return NULL;
+  }
+
+  if (!lfds711_queue_umm_dequeue(&queue->queue, &qe)) {
+    return NULL;
+  }
+
+  p = LFDS711_QUEUE_UMM_GET_VALUE_FROM_ELEMENT(*qe);
+
+  LFDS711_FREELIST_SET_VALUE_IN_ELEMENT(p->freelist_element, p);
```
Comment (th0br0) [inline utils/containers/lock_free/lf_queue.c.tpl]: This shouldn't be necessary each time, we only need to do this once (when malloc'ing this element?)

File: utils/containers/lock_free/lf_queue.c.tpl [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+  struct lfds711_freelist_element* fe;
+  if (lfds711_freelist_pop(&queue->freelist, &fe, NULL)) {
+    p = LFDS711_FREELIST_GET_VALUE_FROM_ELEMENT(*fe);
+  } else if ((p = malloc(sizeof(iota_lf_queue_umm_{TYPE}_t))) == NULL) {
+    return RC_OOM;
+  }
+
+  memcpy(&p->data, data, sizeof(p->data));
+  LFDS711_QUEUE_UMM_SET_VALUE_IN_ELEMENT(p->queue_element, p);
```
Comment (th0br0) [inline utils/containers/lock_free/lf_queue.c.tpl:67]: + `LFDS711_FREELIST_SET_VALUE_IN_ELEMENT(p->freelist_element, p);`

Comment (th0br0) [reply utils/containers/lock_free/lf_queue.c.tpl:67]: also probably should only happen if we had to malloc

File: utils/containers/lock_free/lf_queue.c.tpl [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  LFDS711_FREELIST_SET_VALUE_IN_ELEMENT(p->freelist_element, p);
+  lfds711_freelist_push(&queue->freelist, &p->freelist_element, NULL);
+
+  return p;
+}
+
+size_t iota_lock_free_queue_umm_count(const iota_lf_umm_queue_{TYPE}_t* const queue) {
+  lfds711_pal_uint_t count;
+  lfds711_queue_umm_query(&queue->queue,
+                          LFDS711_QUEUE_UMM_QUERY_SINGLETHREADED_GET_COUNT,
```
Comment (th0br0) [inline utils/containers/lock_free/lf_queue.c.tpl:95]: this is - in theory - very unsafe and can crash, just FYI

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: utils/containers/lock_free/lf_queue.c.tpl (hunk 1 above)
Comment (tsvisabo) [reply utils/containers/lock_free/lf_queue.c.tpl:35]: your'e right

File: utils/containers/lock_free/lf_queue.c.tpl (hunk 4 above)
Comment (tsvisabo) [reply utils/containers/lock_free/lf_queue.c.tpl]: works as well, though it's just pointer assignment which

File: utils/containers/lock_free/lf_queue.c.tpl (hunk 5 above)
Comment (tsvisabo) [reply utils/containers/lock_free/lf_queue.c.tpl:67]: yes

File: utils/containers/lock_free/lf_queue.c.tpl (hunk 6 above)
Comment (tsvisabo) [reply utils/containers/lock_free/lf_queue.c.tpl:95]: which part of it? LFDS711_QUEUE_UMM_QUERY_SINGLETHREADED_GET_COUNT in general?

Comment (th0br0) [reply utils/containers/lock_free/lf_queue.c.tpl:95]: yup

**Review by th0br0 — APPROVED** [review]

**Pull request merged**